### PR TITLE
`Communication`: Send emails for announcement messages

### PR DESCRIPTION
**Pull request by lennart-keller**
**For testing: Deploy on legacy ts1 or ts0 (needs email support)**

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).(https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

With #7437 it's no longer to send announcement posts, but only announcement messages. However, these do not support email notifications yet.

### Description

When writing a message in an announcement channel, an email will be sent to all course members. For announcement channels, a modal opens, where instructors can enter a title and the content for the announcement. Both will be present in the email. The list of messages in the web client also shows the title for announcements.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student where you have access to their email address (e.g. use your TUM account)
- 1 Student that has no admin rights (your TUM account probably has them)
- 1 course where the instructor and the students are a member of

1. Navigate to the messaging page of the course with the instructor
2. Go to the announcement channel and write an announcement. A modal should open.
3. Send the announcement. Verify that the student receives an email. Verify it looks fine and the link works.
4. Visit the announcement channel with the student and verify that you cannot send an announcement as the student.
5. Deactivate email notifications for announcements in the student's setting
6. Write another announcement and verify that the student does not receive an email.

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

For instructors, within an announcement channel, there is a new button to create announcements:
announcement channel

By clicking on this button, a modal opens where the information for the announcement can be entered:
announcement modal

All course members receive an email, if they enabled this in the settings:
email

### CODE REVIEW

**Review by jakubriegel — APPROVED** [review]
Works fine on ts1

**Review by vinceclifford — DISMISSED** [review]
Works as expected. Tested during testing session

**Review by aplr — APPROVED** [review]
Tested in testing session, LGMT

File: src/main/java/de/tum/in/www1/artemis/domain/ConversationNotificationRecipientSummary.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * @param userLogin              the login of the user who is a member of the conversation
+ * @param firstName              the first name of the user
+ * @param lastName               the last name of the user
+ * @param userLangKey            the language key of the user
+ * @param userEmail              the email address of the user
+ * @param isConversationHidden   true if the user has hidden the conversation
+ * @param isAtLeastTutorInCourse true if the user is at least a tutor in the course
+ */
+public record ConversationNotificationRecipientSummary(Long userId, String userLogin, String firstName, String lastName, String userLangKey, String userEmail,
+        boolean isConversationHidden, boolean isAtLeastTutorInCourse) {
```
Comment (tobias-lippert) [inline src/main/java/de/tum/in/www1/artemis/domain/ConversationNotificationRecipientSummary.java:17]: is the user id really nullable?

File: src/test/java/de/tum/in/www1/artemis/metis/MessageIntegrationTest.java [hunk 2]
```diff
@@ -177,23 +180,33 @@ void testCreateConversationPost(CourseInformationSharingConfiguration courseInfo
...
         Post createdPost = request.postWithResponseBody("/api/courses/" + courseId + "/messages", postToSave, Post.class, HttpStatus.CREATED);
         checkCreatedMessagePost(postToSave, createdPost);
 
         // conversation participants should be notified via one broadcast
         verify(websocketMessagingService, never()).sendMessageToUser(anyString(), anyString(), any(PostDTO.class));
         verify(websocketMessagingService, timeout(2000).times(1)).sendMessage(eq("/topic/metis/courses/" + courseId), any(PostDTO.class));
+
+        if (isAnnouncement) {
+            verify(mailService, timeout(2000).times(1)).sendNotification(any(ConversationNotification.class), eq(otherParticipant.getUser()), any(Post.class));
+            verify(mailService, timeout(2000).times(1)).sendNotification(any(ConversationNotification.class), eq(author.getUser()), any(Post.class));
```
Comment (tobias-lippert) [inline src/test/java/de/tum/in/www1/artemis/metis/MessageIntegrationTest.java:205]: Should the author really receive an email?
I always hate these notifications that notify me about an action I've done.

**Review by tobias-lippert — COMMENTED** [review]
Code looks good in general. I've just left two questions.

File: src/main/webapp/app/overview/course-conversations/layout/conversation-messages/conversation-messages.component.html [hunk 3]
```diff
@@ -73,13 +73,22 @@
...
+                    </button>
+                    <jhi-post-create-edit-modal
+                        #createEditModal
+                        [posting]="newPost!"
+                        [isCourseMessagesPage]="true"
+                        (onCreate)="createEmptyPost(); scrollToBottomOfMessages()"
+                    ></jhi-post-create-edit-modal>
+                </div>
+                <ng-template #inlineInput>
+                    <jhi-message-inline-input class="message-input" [posting]="newPost!" (onCreate)="createEmptyPost(); scrollToBottomOfMessages()"> </jhi-message-inline-input>
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/overview/course-conversations/layout/conversation-messages/conversation-messages.component.html:89]:
```suggestion
                    <jhi-message-inline-input class="message-input" [posting]="newPost!" (onCreate)="createEmptyPost(); scrollToBottomOfMessages()" />

File: src/main/webapp/app/overview/course-conversations/other/channel-icon/channel-icon.component.html [hunk 4]
```diff
@@ -2,5 +2,5 @@
     <fa-icon *ngIf="isArchived === true" [icon]="faBoxArchive" size="xs"></fa-icon>
     <fa-icon *ngIf="isPublic === true" [icon]="faHashtag" size="xs"></fa-icon>
     <fa-icon *ngIf="isPublic === false" [icon]="faLock" size="xs"></fa-icon>
-    <fa-icon *ngIf="isAnnouncementChannel === true" [icon]="faBell" size="xs"></fa-icon>
+    <fa-icon *ngIf="isAnnouncementChannel === true" [icon]="faBullhorn" size="xs"></fa-icon>
```
Comment (MaximilianAnzinger) [inline src/main/webapp/app/overview/course-conversations/other/channel-icon/channel-icon.component.html:5]:
```suggestion
    <fa-icon *ngIf="isAnnouncementChannel" [icon]="faBullhorn" size="xs"></fa-icon>
```

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]
Left some comments

File: src/test/java/de/tum/in/www1/artemis/metis/MessageIntegrationTest.java (hunk 2 above)
Comment (lennart-keller) [reply src/test/java/de/tum/in/www1/artemis/metis/MessageIntegrationTest.java:205]: Yes, the author should receive an email. However, I added a check for the email settings of the author, so they only get notified if they have email notifications enabled for the notification type.

File: src/main/java/de/tum/in/www1/artemis/domain/ConversationNotificationRecipientSummary.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    public ConversationNotificationRecipientSummary(Long userId, String userLogin, String firstName, String lastName, String userLangKey, String userEmail,
```
[the remark this answers is not in the document]
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/domain/ConversationNotificationRecipientSummary.java:19]: Yes, I removed it.

File: src/main/java/de/tum/in/www1/artemis/domain/ConversationNotificationRecipientSummary.java (hunk 1 above)
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/domain/ConversationNotificationRecipientSummary.java:17]: In practice it is not. I changed the type to "long".

**Review by vinceclifford — APPROVED** [review]
Reapproved

**Review by aplr — APPROVED** [review]

**Review by tobias-lippert — APPROVED** [review]
Code

**Review by jakubriegel — APPROVED** [review]
Email received, code ok

**Review by MaximilianAnzinger — APPROVED** [review]
Thx for considering my suggestions. Code lgtm 🚀

**Pull request merged**